### ISSUE TITLE
Random tokens being sampled from fp16 models

### ISSUE THREAD

**Issue by finetuneanon** [cloveranon/Clover-Edition#168]
Recently noticed an issue where random tokens could be sampled from fp16 models. After checking the logits, turned out torch.multinomial will sometimes pick zero probability tokens. Logits in this screenshot had all values except four set to -inf, was then F.softmaxed so all values except the four given ones are zero.

![multinomial](https://user-images.githubusercontent.com/82650881/117983828-99cf3980-b337-11eb-8cef-14925f0d4bf8.png)

There's an easy workaround. Just cast logits to fp32, changing the code here:

https://github.com/cloveranon/Clover-Edition/blob/master/aidungeon/gpt2generator.py#L177-L179

To:

```
                next_token = torch.multinomial(
                    F.softmax(logits, dim=-1).float(), num_samples=1
                )
```

**Comment by cloveranon** [comment]
I'm having trouble replicating this. What a terrifying bug.
`a=torch.zeros(50000).to('cuda').to(torch.float16)
a[100]=0.001
sum(torch.multinomial(a, replacement=True, num_samples=1000000)!=100).item()
#should return 0`

It may be worth converting the logits to 32 bits *before* the softmax for numerical accuracy anyway.

**Comment by cloveranon** [comment]
Possibly related https://github.com/pytorch/pytorch/issues/13867

**Comment by finetuneanon** [comment]
It doesn't seem to happen with replacement=True. This code should replicate the issue:

```
import torch
from torch.nn import functional as F
a=torch.zeros(50000).to('cuda').to(torch.float16)
a[100] = b
for i in range(5000):
  choice = torch.multinomial(a, num_samples=1)
  if choice.item() != 100:
    print(choice)
```

I agree that converting the logits to fp32 before softmax is a good idea.

**Comment by cloveranon** [comment]
I believe this is already in the code here: https://github.com/cloveranon/Clover-Edition/blame/400a73e49378bd41e476cdba89e4aa2be10ae318/gpt2generator.py#L128

This seems like an incredibly serious bug, there should at least be a warning if they can't fix it for whatever reason.

**Comment by finetuneanon** [comment]
Missed that, never mind then, thanks.
